File: benchmarks/runners/node/coldstart-child.mjs
// Fresh-process worker for one case. Prints a "READY" marker the instant the
// runtime is initialized, then a JSON line with the in-process first-run timing.
// The PARENT (coldstart.mjs) measures cold-start as wall-clock from spawn to the
// READY marker, so process launch + module/addon load + isolate init are all
// included — not just the in-process initialize() call.
import { readFileSync, writeSync } from "node:fs";
import { join } from "node:path";
import { loadManifest, resolveInputs } from "../../lib/manifest.mjs";
import { loadWrapper } from "./wrapper.mjs";

const [, , corpusDir, caseId] = process.argv;
const nowNs = () => process.hrtime.bigint();
const msSince = (s) => Number(nowNs() - s) / 1e6;

const manifest = loadManifest(corpusDir);
const c = manifest.cases.find((x) => x.id === caseId);
if (!c) throw new Error(`unknown case: ${caseId}`);
const script = readFileSync(join(corpusDir, c.script), "utf-8");

const resolved = resolveInputs(manifest, c);
const inputs = {};
for (const [name, v] of Object.entries(resolved)) {
  inputs[name] = { content: v.buffer, mimeType: v.mimeType, charset: v.charset ?? "utf-8" };
}

const api = await loadWrapper();
const dw = new api.DataWeave();

dw.initialize();
// Runtime is ready: flush the marker synchronously so the parent's clock stops
// here. writeSync (fd 1) bypasses the async stdout buffer, so the timestamp the
// parent reads reflects init completion, not stream flushing.
writeSync(1, "READY\n");

const runStart = nowNs();
const result = dw.run(script, inputs);
const firstRunMs = msSince(runStart);
if (!result.success) throw new Error(`first run failed: ${result.error}`);

dw.cleanup();
process.stdout.write(JSON.stringify({ firstRunMs }) + "\n");

File: benchmarks/runners/python/coldstart_child.py
"""Fresh-process worker for one case. Prints a "READY" marker the instant the
runtime is initialized, then a JSON line with the in-process first-run timing.
The PARENT (coldstart.py) measures cold-start as wall-clock from spawn to the
READY marker, so process launch + interpreter start + dlopen(dwlib) + isolate
creation are all included — not just the in-process initialize() call."""

import json
import sys
import time

from manifest import load_manifest, read_script, resolve_inputs
from wrapper import load_wrapper


def _to_inputs(resolved):
    inputs = {}
    for name, v in resolved.items():
        inputs[name] = {
            "content": v["bytes"],
            "mimeType": v["mimeType"],
            "charset": v["charset"] or "utf-8",
        }
    return inputs


def main(argv):
    corpus_dir, case_id = argv[1], argv[2]
    manifest = load_manifest(corpus_dir)
    c = next((x for x in manifest["cases"] if x["id"] == case_id), None)
    if c is None:
        raise SystemExit(f"unknown case: {case_id}")
    script = read_script(manifest, c)
    inputs = _to_inputs(resolve_inputs(manifest, c))

    api = load_wrapper()
    dw = api.DataWeave()

    dw.initialize()
    # Runtime is ready: flush the marker now so the parent's clock stops here.
    # flush() guarantees the byte is on the pipe before the timed run begins.
    sys.stdout.write("READY\n")
    sys.stdout.flush()

    run_start = time.perf_counter_ns()
    result = dw.run(script, inputs)
    first_run_ms = (time.perf_counter_ns() - run_start) / 1e6
    if not result.success:
        raise SystemExit(f"first run failed: {result.error}")

    dw.cleanup()
    sys.stdout.write(json.dumps({"firstRunMs": first_run_ms}) + "\n")


if __name__ == "__main__":
    main(sys.argv)
